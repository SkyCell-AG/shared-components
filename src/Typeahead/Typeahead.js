import React, {
    useCallback,
    useMemo,
} from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import noop from 'lodash/noop'
import omit from 'lodash/omit'
import delay from 'lodash/delay'

import InputSingleLine from 'InputSingleLine'
import WithStatusHandler from 'WithStatusHandler'

import useStyles from './Typeahead.styles'

const propTypes = {
    searchPhrase: PropTypes.string.isRequired,
    name: PropTypes.string,
    value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
    onChange: PropTypes.func,
    title: PropTypes.string,
    setSearchPhrase: PropTypes.func.isRequired,
    options: (props, propName, componentName) => {
        const {
            options,
            labelField,
            valueField,
        } = props

        if (!Array.isArray(options)) {
            return new Error(`Invalid prop ${propName} in Component ${componentName} need to be an array`)
        }

        const notValidElementIndex = options.findIndex((option) => {
            return !(option[labelField] && option[valueField])
        })

        if (notValidElementIndex !== -1) {
            return new Error(`Invalid element with index ${notValidElementIndex} in prop ${propName} in Component ${componentName}.`)
        }

        return null
    },
    optionsVisible: PropTypes.bool.isRequired,
    labelField: PropTypes.string.isRequired,
    valueField: PropTypes.string.isRequired,
    openOptions: PropTypes.func.isRequired,
    closeOptions: PropTypes.func.isRequired,
    variant: PropTypes.oneOf([
        'standard',
        'outlined',
        'filled',
    ]),
    classNames: PropTypes.shape({
        options: PropTypes.string,
    }),
    className: PropTypes.string,
}

const defaultProps = {
    value: null,
    name: '',
    onChange: noop,
    options: [],
    title: '',
    variant: 'outlined',
    classNames: {},
    className: '',
}

const removeFields = (obj) => {
    return omit(obj, [
        'value',
        'Comp',
        'valueField',
        'labelField',
        'loadLabel',
        'loadOptions',
        'options',
        'loadAirports',
        'className',
    ])
}
const Typeahead = (props) => {
    const {
        onChange,
        options,
        openOptions,
        optionsVisible,
        setSearchPhrase,
        closeOptions,
        searchPhrase,
        valueField,
        labelField,
        variant,
        value,
        name,
        classNames,
        className,
    } = props

    const classes = useStyles()

    const onOptionClick = useCallback((event) => {
        const {
            target: {
                dataset: {
                    value: newValue,
                },
            },
        } = event

        onChange(newValue, {
            target: {
                name,
                value: newValue,
            },
        })
        closeOptions()
    }, [
        closeOptions,
        name,
        onChange,
    ])

    const onEnterPress = useCallback((event) => {
        const {
            key,
        } = event

        if (optionsVisible && key === 'Enter') {
            onChange(options[0].id, {
                target: {
                    name,
                    value: options[0].id,
                },
            })
            closeOptions()
        }
    }, [
        closeOptions,
        name,
        onChange,
        options,
        optionsVisible,
    ])

    const onBlur = useMemo(() => {
        return () => {
            return delay(closeOptions, 300)
        }
    }, [closeOptions])

    return (
        <div
            className={clsx(classes.root, className)}
        >
            <InputSingleLine
                {...removeFields(props)}
                onChange={setSearchPhrase}
                value={optionsVisible ? searchPhrase : value}
                onKeyPress={onEnterPress}
                onFocus={openOptions}
                onBlur={onBlur}
                variant={variant}
                autoComplete="off"
            />
            <div className={
                clsx(
                    classes.optionsWrapper,
                    optionsVisible && 'active',
                )
            }
            >
                {options.length > 0 && (
                    <div className={clsx(classes.options, classNames.options)}>
                        {options.map((option) => {
                            return (
                                <button
                                    type="button"
                                    key={`option-${option[valueField]}`}
                                    onClick={onOptionClick}
                                    className={classes.option}
                                    data-value={option[valueField]}
                                >
                                    {option[labelField]}
                                </button>
                            )
                        })}
                    </div>
                )}
            </div>
        </div>
    )
}

Typeahead.propTypes = propTypes
Typeahead.defaultProps = defaultProps

export default WithStatusHandler(Typeahead)
