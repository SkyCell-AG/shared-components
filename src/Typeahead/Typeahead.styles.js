import {
    makeStyles,
} from '@material-ui/core'

const useStyles = makeStyles((theme) => {
    return {
        root: {},
        optionsWrapper: {
            zIndex: 1000,
            position: 'relative',
            display: 'none',
            '&.active': {
                display: 'block',
            },
        },
        options: {
            position: 'absolute',
            display: 'flex',
            flexDirection: 'column',
            width: '100%',
            border: `1px solid ${theme.palette.secondary[200]}`,
            background: theme.palette.common.white,
            height: 200,
            overflow: 'auto',
        },
        option: {
            textAlign: 'left',
            cursor: 'pointer',
            border: 'none',
            background: 'none',
            fontWeight: theme.typography.body1.fontWeight,
            lineHeight: theme.typography.body1.lineHeight,
            color: theme.typography.body1.color,
            fontSize: theme.typography.body1.fontSize,
            padding: theme.spacing(1),
            '&:hover': {
                background: theme.palette.primary[30],
            },
        },
    }
})

export default useStyles
