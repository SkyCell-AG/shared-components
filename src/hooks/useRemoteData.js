import {
    useReducer,
    useCallback,
    useEffect,
    useRef,
} from 'react'

import generateAsyncActions from 'utils/generateAsyncActions'
import createReducer from 'utils/createReducer'
import createCancelablePromise from 'utils/createCancelablePromise'
import {
    PRISTIN,
    PENDING,
    SUCCESS,
    FAILURE,
} from 'utils/requestStatuses'

const LOAD = generateAsyncActions('LOAD')

const initState = {
    data: undefined,
    err: undefined,
    status: PRISTIN,
}

const reducer = createReducer({
    [LOAD.pending]: (state) => {
        return {
            ...state,
            status: PENDING,
        }
    },
    [LOAD.success]: (state, {
        payload: {
            data,
        },
    }) => {
        return {
            ...state,
            status: SUCCESS,
            data,
        }
    },
    [LOAD.failure]: (state, {
        err,
    }) => {
        return {
            ...state,
            status: FAILURE,
            err,
        }
    },
})

const useRemoteData = (
    load,
    wantToLoad = false,
) => {
    const [
        state,
        dispatch,
    ] = useReducer(reducer, initState)

    const cancelRequest = useRef(null)

    const loadData = useCallback((...params) => {
        dispatch({
            type: LOAD.pending,
        })

        const [
            request,
            cancel,
        ] = createCancelablePromise(load(...params))

        cancelRequest.current = cancel

        request.then((data) => {
            cancelRequest.current = null
            dispatch({
                type: LOAD.success,
                payload: {
                    data,
                },
            })
        }).catch((err) => {
            if (err.message === 'canceled') {
                return
            }

            dispatch({
                type: LOAD.failure,
                err,
            })
        })
    }, [
        dispatch,
        load,
    ])

    useEffect(() => {
        if (cancelRequest.current) {
            cancelRequest.current()
        }
        return () => {
            if (cancelRequest.current) {
                cancelRequest.current()
            }
        }
    }, [loadData])

    useEffect(() => {
        if (wantToLoad) {
            loadData()
        }
    }, [
        loadData,
        wantToLoad,
    ])

    return [
        state,
        loadData,
    ]
}

export default useRemoteData
