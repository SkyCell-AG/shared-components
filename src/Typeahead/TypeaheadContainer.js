import React, {
    useReducer,
    useEffect,
    useCallback,
} from 'react'
import PropTypes from 'prop-types'
import noop from 'lodash/noop'

import useRemoteData from 'hooks/useRemoteData'
import createReducer from 'utils/createReducer'

import Typeahead from './Typeahead'

const propTypes = {
    loadOptions: PropTypes.func.isRequired,
    loadLabel: PropTypes.func,
    value: PropTypes.string,
    valueField: PropTypes.string,
    labelField: PropTypes.string,
}
const defaultProps = {
    valueField: 'id',
    loadLabel: noop,
    value: '',
    labelField: '',
}

const OPEN_OPTIONS = 'OPEN_OPTIONS'
const CLOSE_OPTIONS = 'CLOSE_OPTIONS'
const SET_SEARCH_PHRASE = 'SET_SEARCH_PHRASE'

const reducer = createReducer({
    [SET_SEARCH_PHRASE]: (state, {
        searchPhrase,
    }) => {
        return {
            ...state,
            searchPhrase,
        }
    },
    [OPEN_OPTIONS]: (state) => {
        return {
            ...state,
            optionsVisible: true,
            searchPhrase: '',
        }
    },
    [CLOSE_OPTIONS]: (state) => {
        return {
            ...state,
            optionsVisible: false,
            searchPhrase: '',
        }
    },
})

const TypeaheadContainer = (props) => {
    const {
        loadOptions,
        loadLabel,
        value,
        labelField,
    } = props

    const [
        {
            optionsVisible,
            searchPhrase,
        },
        dispatch,
    ] = useReducer(reducer, {
        optionsVisible: false,
        searchPhrase: '',
    })

    const openOptions = useCallback(() => {
        dispatch({
            type: OPEN_OPTIONS,
        })
    }, [dispatch])

    const closeOptions = useCallback(() => {
        dispatch({
            type: CLOSE_OPTIONS,
        })
    }, [dispatch])

    const setSearchPhrase = useCallback((newSearchPhrase) => {
        dispatch({
            type: SET_SEARCH_PHRASE,
            searchPhrase: newSearchPhrase,
        })
    }, [dispatch])

    const [
        {
            data: options,
        },
        loadData,
    ] = useRemoteData(loadOptions)

    useEffect(() => {
        if (searchPhrase) {
            loadData(searchPhrase)
        }
    }, [
        loadData,
        searchPhrase,
    ])

    const getLabel = useCallback(() => {
        if (!value) {
            return Promise.resolve('')
        }
        const labels = loadLabel(value)

        if (!labels) {
            return Promise.resolve(value)
        }

        return labels.then(({
            data,
        }) => {
            const label = data[0][labelField]

            return label
        })
    }, [
        labelField,
        loadLabel,
        value,
    ])

    const [state] = useRemoteData(getLabel, true)

    return (
        <Typeahead
            {...props}
            {...state}
            value={state.data}
            setSearchPhrase={setSearchPhrase}
            searchPhrase={searchPhrase}
            options={options}
            closeOptions={closeOptions}
            optionsVisible={optionsVisible}
            openOptions={openOptions}
        />
    )
}

TypeaheadContainer.propTypes = propTypes
TypeaheadContainer.defaultProps = defaultProps

export default TypeaheadContainer
