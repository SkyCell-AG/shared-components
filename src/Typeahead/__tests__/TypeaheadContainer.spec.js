import React from 'react'
import {
    mount,
} from 'enzyme'

import useRemoteData from 'hooks/useRemoteData'

import TypeaheadContainer from '../TypeaheadContainer'
import Typeahead from '../Typeahead'

jest.mock('hooks/useRemoteData')
jest.mock('../Typeahead', () => {
    return () => {
        return <div>empty</div>
    }
})

describe('TypeaheadContainer', () => {
    it('child props', () => {
        const loadLabel = jest.fn().mockResolvedValue({
            data: [{
                id: 'value-id',
                label: 'value-label',
            }],
        })

        const loadOptions = jest.fn().mockResolvedValue({
            data: [
                {
                    id: 'id1',
                    label: 'label1',
                    field: 'field',
                },
                {
                    id: 'id2',
                    label: 'label2',
                },
            ],
        })

        const loadOptionsBySearchPhrase = jest.fn()

        useRemoteData
            .mockReturnValueOnce([
                {
                    data: ['option'],
                },
                loadOptionsBySearchPhrase,
            ])
            .mockReturnValueOnce([{
                state: 'of load label',
            }])

        const wrapper = mount(<TypeaheadContainer
            labelField="label"
            valueField="id"
            loadOptions={loadOptions}
            loadLabel={loadLabel}
            value="some value"
        />)

        const child = wrapper.find(Typeahead)

        expect(child.props()).toMatchSnapshot()
    })
})
