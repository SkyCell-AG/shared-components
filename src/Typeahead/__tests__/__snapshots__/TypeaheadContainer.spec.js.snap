// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeaheadContainer child props 1`] = `
Object {
  "closeOptions": [Function],
  "labelField": "label",
  "loadLabel": [MockFunction],
  "loadOptions": [MockFunction],
  "openOptions": [Function],
  "options": Array [
    "option",
  ],
  "optionsVisible": false,
  "searchPhrase": "",
  "setSearchPhrase": [Function],
  "state": "of load label",
  "value": undefined,
  "valueField": "id",
}
`;
